File: exercises/py110/easy2/cute_angles.py

'''
PEDAC
Input:  a floating point number representing an angle between 0 and 360 degrees. 
Output: a string representing the angle in degrees, minutes, and seconds. 

Rules
Explicit:

Implicit:
    - 360 degrees = 0 degrees. If the angle exceeds 360 deg, it will be original degree - 360 degree.
    - Degree: whole number to the left of the decimal point
    - Minute: Take decimal part of the degrees and multiply it by 60
        - The whole number is your minutes
    - Seconds: take the remaining decimal part from previous step and multiply by 60
        - the result is seconds. Round to the nearest whole number if there is a decimal remaining
    
Questions:

Data Structures:
    - list: to store degree, minute, and seconds. 
    - float: involves multiplication and divmod
    - string: to make the final string that is returned

Algorithms:
    - Given the original floating point number "number", get a quotient and modulus of that number by 
        dividing the number by 1
            -"degree" variable: quotient
            -"remainder" variable: modulus part
    -We use "remainder" variable from the previous step to get "minute"
        - Multiply "remainder" by 60. 
        - Whole number part is "minute"
        - Store decimals as "remainder"
    -We use "remainder" variable from the previous step to get "second"
        - Multiply "remainder" by 60.
        - Whole number part is "second".
    - Compose and return a string with "degree", "remainder", and "second" variables using 
      string interpolation
        - f-string

Challenge 2
Degree not 0 <= angle <= 360 
Algorithm:
'''

DEGREE = "\u00B0" # degree symbol
def dms(angle):
    is_negative = False 
    if angle < 0:
        is_negative = True 

    angle = abs(angle)
    degree, remainder = divmod(angle, 1)
    remainder = remainder * 60
    minute, remainder = divmod(remainder, 1)
    remainder = remainder * 60
    second, remainder = divmod(remainder, 1)

    if is_negative:
        return f"-{int(degree)}{DEGREE}{int(minute):02d}'{int(second):02d}\""
    else:
        return f"{int(degree)}{DEGREE}{int(minute):02d}'{int(second):02d}\""

# All of these examples should print True
print(dms(30) == "30°00'00\"")
print(dms(76.73) == "76°43'48\"")
print(dms(254.6) == "254°35'59\"" or dms(254.6) == "254°36'00\"")
print(dms(93.034773) == "93°02'05\"")
print(dms(0) == "0°00'00\"")
print(dms(360) == "360°00'00\"" or dms(360) == "0°00'00\"")

# Challenge 1 (Got'em)
# print(dms(-1))   # -1°00'00"
# print(dms(400))  # 400°00'00"
# print(dms(-40))  # -40°00'00"
# print(dms(-420)) # -420°00'00"

# Challenge 2
print(dms(-1))   # 359°00'00"
print(dms(400))  # 40°00'00"
print(dms(-40))  # 320°00'00"
print(dms(-420)) # 300°00'00"

File: exercises/LS215/easy4/cute-angles.js
/*

=== Problem ===
Input: floating point number
        - representing an angle between 0 and 360 degrees
Output: a string
        - representing that angle in degrees, minutes, and seconds

Rules:
  - use a degree symbol (°) to represent degrees, 
  - a single quote (') to represent minutes, and 
  - a double quote (") to represent seconds.

  - There are 60 minutes in a degree, and 
  - there are 60 seconds in a minute.

=== Examples === 
30 => "30°00'00\""
76.73 => "76°43'48\""
0 => "0°00'00\""
254.6 => "254°35'59\""
360 => "360°00'00\"" || "0°00'00\""

=== Questions === 
- What if angle is not an int
  - angle will be an int 
- What if angle is a number less than 0
  - angle will always be an int greater than or equal to 0

=== Data Structures ===
- Number (int, float)
  - decimal = angle % 1
  - 60 mins in an angle
    - 1 angle = 60 mins
    - 1 / 60 angle = min 
    - 0.5 => what minute? 
    - 0.5 angle => 30 mins 
    - angle * 60
  - 60 sec in a min
    - 1 min = 60 sec

=== Algo === 
- Filter out whole number and decimal
  - wholeNumber = Math.floor(angle / 1)
  - decimal = angle % 1


*/

const DEGREE = '\u00B0'; // degree symbol
// console.log(Math.floor(76.73 / 1));
// console.log(Math.floor(360 / 1));
// console.log(Math.floor(0 / 1));
// console.log(Math.floor(93.034773 / 1));

function dms(angle) {
  let decimal = angle % 1;
  let wholeNumber = Math.floor(angle / 1);
  // console.log(String(decimal * 60).padStart(0, 2));
  // console.log(decimal * 60);
  // console.log(String(Math.floor(decimal * 60)).padStart(2, 0));
  let minute = Math.floor(decimal * 60);
  let second = Math.floor(((decimal * 60) % 1) * 60);

  console.log(`${String(wholeNumber)}${DEGREE}${String(minute).padStart(2, 0)}'${String(second).padStart(2, 0)}"`);
  return `${String(wholeNumber)}${DEGREE}${String(minute).padStart(2, 0)}'${String(second).padStart(2, 0)}"`
}

// All test cases should return true
console.log(dms(30) === "30°00'00\"");
console.log(dms(76.73) === "76°43'48\"");
console.log(dms(254.6) === "254°35'59\"");
console.log(dms(93.034773) === "93°02'05\"");
console.log(dms(0) === "0°00'00\"");
console.log(dms(360) === "360°00'00\"" || dms(360) === "0°00'00\"");